activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -193,6 +205,80 @@ def wait_poll(timeout)
         end
       end
 
+      # Adds the ability to turn a basic fair FIFO queue into one
+      # biased to some thread.
+      module BiasableQueue # :nodoc:
+        class BiasedConditionVariable
```

Comment:
@jdantonio thanks for cc-ing me. I'll have some time on weekend, I'll look in more detail.